File: solutions/go/need-for-speed/1/need_for_speed.go
package speed

// TODO: define the 'Car' type struct
type Car struct {
    battery, batteryDrain, speed, distance int
}

// NewCar creates a new remote controlled car with full battery and given specifications.
func NewCar(speed, batteryDrain int) Car {
	return Car {speed: speed, batteryDrain: batteryDrain, battery: 100, distance: 0}
}

// TODO: define the 'Track' type struct
type Track struct {
    distance int
}

// NewTrack creates a new track
func NewTrack(distance int) Track {
	return Track {distance: distance}
}

// Drive drives the car one time. If there is not enough battery to drive one more time,
// the car will not move.
func Drive(car Car) Car {
    if car.battery < car.batteryDrain {
        return car
    }
	car.battery -= car.batteryDrain
    car.distance += car.speed
    return car
}

// CanFinish checks if a car is able to finish a certain track.
func CanFinish(car Car, track Track) bool {
	return track.distance <= car.speed*car.battery/car.batteryDrain
}

File: solutions/java/need-for-speed/1/src/main/java/NeedForSpeed.java
final class NeedForSpeed {
    private final int speed;
    private final int batteryDrain;
    private int distanceDriven = 0;
    private int batteryRemaining = 100;
    
    NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public boolean batteryDrained() {
        return batteryRemaining < batteryDrain;
    }

    public int distanceDriven() {
        return distanceDriven;
    }

    public void drive() {
        if (batteryDrained()) {
            return;
        }
        distanceDriven += speed;
        batteryRemaining -= batteryDrain;
    }

    public int maximumDistance() {
        return batteryRemaining / batteryDrain * speed;
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

final class RaceTrack {
    private final int distance;
    
    RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean canFinishRace(NeedForSpeed car) {
        return car.maximumDistance() >= distance;
    }
}
